clippy_lints/src/default_trait_access.rs

```diff
@@ -0,0 +1,68 @@
+use rustc::hir::*;
+use rustc::lint::*;
+use rustc::ty::TypeVariants;
+
+use crate::utils::{any_parent_is_automatically_derived, match_def_path, opt_def_id, paths, span_lint_and_sugg};
+
+
+/// **What it does:** Checks for literal calls to `Default::default()`.
+///
+/// **Why is this bad?** It's more clear to the reader to use the name of the type whose default is
+/// being gotten than the generic `Default`.
+///
+/// **Known problems:** None.
+///
+/// **Example:**
+/// ```rust
+/// // Bad
+/// let s: String = Default::default();
+///
+/// // Good
+/// let s = String::default();
+/// ```
+declare_clippy_lint! {
+    pub DEFAULT_TRAIT_ACCESS,
+    style,
+    "checks for literal calls to Default::default()"
+}
+
+#[derive(Copy, Clone)]
+pub struct DefaultTraitAccess;
+
+impl LintPass for DefaultTraitAccess {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(DEFAULT_TRAIT_ACCESS)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for DefaultTraitAccess {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if_chain! {
+            if let ExprCall(ref path, ..) = expr.node;
+            if !any_parent_is_automatically_derived(cx.tcx, expr.id);
```

Comment:
You found a workaround for the currently broken `in_macro`! Sweet.

@rust-lang-nursery/clippy  do we want to add this to the in_macro check? It's probably expensive to do, but we could monoize it via a thread local HashMap

Reply:
I should have formulated this better:

`in_macro` is broken and does not help us anymore in detecting `derive`s. You just found a way to detect derives again, and we should make `in_macro` also call your function in order to make it work again.

Reply:
Fabulous! :D